lib/rucio/tests/test_download.py

```diff
@@ -463,3 +463,37 @@ def __init__(self, *args, mocks_get=mocks_get, **kwargs):
             mocks_get.clear()
             download_client.download_dids([{'did': did_str, 'base_dir': tmp_dir, 'transfer_speed_timeout': 0}])
             mocks_get[0].assert_called_with(ANY, ANY, transfer_timeout=60)
+
+
+def test_download_file_with_impl(rse_factory, did_factory, download_client):
+    impl = 'rucio.rse.protocols.gfalv2.Default'
+    scope = str(did_factory.default_scope)
+    rse, _ = rse_factory.make_rse_with_impl(impl)
+    did = did_factory.upload_test_file(rse)
```

Comment:
This line fails saying RSE cannot be referenced

Reply:
hi :) just got this via @mlassnig
do you have any stacktrace or test output?

Reply:
hello @bziemons,
```
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
s..F.....
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw5] linux -- Python 3.6.8 /usr/bin/python

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f93baeed048>, path = 'test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE'

    def __gfal2_exist(self, path):
        """
        Uses gfal2 to check whether the file exists.
    
        :param path: Physical file name
    
        :returns: 0 if it exists, -1 if it doesn't
    
        :raises RucioException: if the error is not source not found.
        """
        ctx = self.__ctx
        try:
>           if ctx.stat(str(path)):
E           gfal2.GError: Protocol not supported or path/url invalid: test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE

../rse/protocols/gfal.py:494: GError

During handling of the above exception, another exception occurred:

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f93baeed048>, path = 'test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE'

    def exists(self, path):
        """
        Checks if the requested file is known by the referred RSE.
    
        :param path: Physical file name
    
        :returns: True if the file exists, False if it doesn't
    
        :raises SourceNotFound: if the source file was not found on the referred storage.
        """
        self.logger(logging.DEBUG, 'checking if file exists {}'.format(path))
    
        try:
>           status = self.__gfal2_exist(path)

../rse/protocols/gfal.py:332: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f93baeed048>, path = 'test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE'

    def __gfal2_exist(self, path):
        """
        Uses gfal2 to check whether the file exists.
    
        :param path: Physical file name
    
        :returns: 0 if it exists, -1 if it doesn't
    
        :raises RucioException: if the error is not source not found.
        """
        ctx = self.__ctx
        try:
            if ctx.stat(str(path)):
                return 0
            return -1
        except gfal2.GError as error:  # pylint: disable=no-member
            if error.code == errno.ENOENT or 'No such file' in str(error):  # pylint: disable=no-member
                return -1
>           raise exception.RucioException(error)
E           rucio.common.exception.RucioException: An unknown exception occurred.
E           Details: Protocol not supported or path/url invalid: test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE

../rse/protocols/gfal.py:500: RucioException

During handling of the above exception, another exception occurred:

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7f93bd6a97b8>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7f93bd6a99e8>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7f93bd6b5a58>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.gfalv2.Default'
        scope = str(did_factory.default_scope)
        rse, _ = rse_factory.make_rse_with_impl(impl)
>       did = did_factory.upload_test_file(rse)

test_download.py:472: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
temp_factories.py:283: in upload_test_file
    self.upload_client.upload([item])
../client/uploadclient.py:237: in upload
    if rsemgr.exists(rse_settings, pfn if pfn else file_did, domain=domain, auth_token=self.auth_token, logger=logger):
../rse/rsemanager.py:302: in exists
    exists = protocol.exists(pfn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f93baeed048>, path = 'test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE'

    def exists(self, path):
        """
        Checks if the requested file is known by the referred RSE.
    
        :param path: Physical file name
    
        :returns: True if the file exists, False if it doesn't
    
        :raises SourceNotFound: if the source file was not found on the referred storage.
        """
        self.logger(logging.DEBUG, 'checking if file exists {}'.format(path))
    
        try:
            status = self.__gfal2_exist(path)
            if status:
                return False
            return True
        except exception.SourceNotFound:
            return False
        except Exception as error:
>           raise exception.ServiceUnavailable(error)
E           rucio.common.exception.ServiceUnavailable: The requested service is not available at the moment.
E           Details: An unknown exception occurred.
E           Details: Protocol not supported or path/url invalid: test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE

../rse/protocols/gfal.py:339: ServiceUnavailable
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
---------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------
WARNING  root:request.py:727 Tried to cancel non-existant request for DID mock:file_ZHPZQYZSXE at RSE MOCK_LMYNMBUIAU
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_blacklisted_replicas
test_download.py::test_download_without_base_dir
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_norandom_respected
test_download.py::test_transfer_timeout
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - rucio.common.exception.ServiceUnavailable: The requested service is not available at the moment.
============================================================================== 1 failed, 7 passed, 1 skipped, 15 warnings in 12.94s ===============================================================================
[root@b2e68ef33f26 rucio]# 

```

this is the test output

Reply:
> Details: Protocol not supported or path/url invalid: test://d68fa5b6074a46acaa5d1af49e5974a9.cern.ch/test/mock/a1/ae/file_ZHPZQYZSXE

@rak108 that seems to be the important message. I don't see how that correlates with what you said above though:

> This line fails saying RSE cannot be referenced

Reply:
@bziemons Yes Im sorry,  I should have been more clear:
So this is the error that crops up with the gfalv2 protocol path.
So the other file that I have edited  contains functions to make RSEs with xrootd protocol, or srm 
So initially after I kept getting this error of protocol path invalid,I decided to try the same thing with the 'rucio.rse.protocols.xrootd.Default', and then it kept giving RSE not referenced. The same goes for srm too

The only situation in which this line would not throw any error whatsoever was if it was run with POSIX, but that (if I'm not mistaken) defeats the purpose as we require testing the download/upload with the RSE that supports the impl protocol as well right

Reply:
ah, wait

`Details: Protocol not supported or path/url invalid: test....`

GFAL requires a protocol that it "understands", so must be one of the installed gfal plugins (srm, gsiftp, root, http, davs, ...) I think there might be a `mock://` one, but there certainly is no `test://` one

Reply:
@mlassnig Oh alright, so i changed the scheme to trying with both MOCK and mock, and it gave me this error (same for both):
```
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
s..F.....
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw4] linux -- Python 3.6.8 /usr/bin/python

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f24020b9198>, path = 'mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY'

    def __gfal2_exist(self, path):
        """
        Uses gfal2 to check whether the file exists.
    
        :param path: Physical file name
    
        :returns: 0 if it exists, -1 if it doesn't
    
        :raises RucioException: if the error is not source not found.
        """
        ctx = self.__ctx
        try:
>           if ctx.stat(str(path)):
E           gfal2.GError: Protocol not supported or path/url invalid: mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY

../rse/protocols/gfal.py:494: GError

During handling of the above exception, another exception occurred:

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f24020b9198>, path = 'mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY'

    def exists(self, path):
        """
        Checks if the requested file is known by the referred RSE.
    
        :param path: Physical file name
    
        :returns: True if the file exists, False if it doesn't
    
        :raises SourceNotFound: if the source file was not found on the referred storage.
        """
        self.logger(logging.DEBUG, 'checking if file exists {}'.format(path))
    
        try:
>           status = self.__gfal2_exist(path)

../rse/protocols/gfal.py:332: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f24020b9198>, path = 'mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY'

    def __gfal2_exist(self, path):
        """
        Uses gfal2 to check whether the file exists.
    
        :param path: Physical file name
    
        :returns: 0 if it exists, -1 if it doesn't
    
        :raises RucioException: if the error is not source not found.
        """
        ctx = self.__ctx
        try:
            if ctx.stat(str(path)):
                return 0
            return -1
        except gfal2.GError as error:  # pylint: disable=no-member
            if error.code == errno.ENOENT or 'No such file' in str(error):  # pylint: disable=no-member
                return -1
>           raise exception.RucioException(error)
E           rucio.common.exception.RucioException: An unknown exception occurred.
E           Details: Protocol not supported or path/url invalid: mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY

../rse/protocols/gfal.py:500: RucioException

During handling of the above exception, another exception occurred:

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7f24048767b8>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7f24048769e8>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7f2404882a58>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.gfalv2.Default'
        scope = str(did_factory.default_scope)
        rse, _ = rse_factory.make_rse_with_impl(impl)
>       did = did_factory.upload_test_file(rse)

test_download.py:472: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
temp_factories.py:283: in upload_test_file
    self.upload_client.upload([item])
../client/uploadclient.py:237: in upload
    if rsemgr.exists(rse_settings, pfn if pfn else file_did, domain=domain, auth_token=self.auth_token, logger=logger):
../rse/rsemanager.py:302: in exists
    exists = protocol.exists(pfn)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.gfalv2.Default object at 0x7f24020b9198>, path = 'mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY'

    def exists(self, path):
        """
        Checks if the requested file is known by the referred RSE.
    
        :param path: Physical file name
    
        :returns: True if the file exists, False if it doesn't
    
        :raises SourceNotFound: if the source file was not found on the referred storage.
        """
        self.logger(logging.DEBUG, 'checking if file exists {}'.format(path))
    
        try:
            status = self.__gfal2_exist(path)
            if status:
                return False
            return True
        except exception.SourceNotFound:
            return False
        except Exception as error:
>           raise exception.ServiceUnavailable(error)
E           rucio.common.exception.ServiceUnavailable: The requested service is not available at the moment.
E           Details: An unknown exception occurred.
E           Details: Protocol not supported or path/url invalid: mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY

../rse/protocols/gfal.py:339: ServiceUnavailable
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
---------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------
WARNING  root:request.py:727 Tried to cancel non-existant request for DID mock:file_HAWCDMUGBY at RSE MOCK_OWQTFJEYHD
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_without_base_dir
test_download.py::test_download_blacklisted_replicas
test_download.py::test_norandom_respected
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_transfer_timeout
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - rucio.common.exception.ServiceUnavailable: The requested service is not available at the moment.
============================================================================== 1 failed, 7 passed, 1 skipped, 15 warnings in 12.29s ===============================================================================
[root@b2e68ef33f26 rucio]# 
```

Reply:
So if instead of making the RSE like this taking in impl parameter, if Instead directly make an RSE with the xrootd or srm protocol using temp_factories make_xroot_rse or make_srm_rse, it still does not work and gives this error of RSE Access denied (here I tried with the make_xroot_rse function):
```
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
s.F......
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw6] linux -- Python 3.6.8 /usr/bin/python

self = <rucio.rse.protocols.xrootd.Default object at 0x7fd395ca7a58>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
>               raise exception.RSEAccessDenied(err)
E               rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E               Details: [FATAL] Invalid address

../rse/protocols/xrootd.py:198: RSEAccessDenied

During handling of the above exception, another exception occurred:

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7fd3984a7710>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7fd3984a7940>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7fd39845fb00>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.xrootd.Default'
        scope = str(did_factory.default_scope)
        rse, _ = rse_factory.make_rse_with_impl(impl)
>       did = did_factory.upload_test_file(rse)

test_download.py:472: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
temp_factories.py:283: in upload_test_file
    self.upload_client.upload([item])
../client/uploadclient.py:237: in upload
    if rsemgr.exists(rse_settings, pfn if pfn else file_did, domain=domain, auth_token=self.auth_token, logger=logger):
../rse/rsemanager.py:279: in exists
    protocol.connect()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.xrootd.Default object at 0x7fd395ca7a58>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
                raise exception.RSEAccessDenied(err)
        except Exception as e:
>           raise exception.RSEAccessDenied(e)
E           rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E           Details: Referenced RSE not reachable.
E           Details: [FATAL] Invalid address

../rse/protocols/xrootd.py:200: RSEAccessDenied
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
---------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------
WARNING  root:request.py:727 Tried to cancel non-existant request for DID mock:file_AUXFBWCIMK at RSE MOCK_RFXNRCTXDT
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_blacklisted_replicas
test_download.py::test_download_without_base_dir
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_transfer_timeout
test_download.py::test_norandom_respected
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
============================================================================== 1 failed, 7 passed, 1 skipped, 15 warnings in 12.64s ===============================================================================
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
sF.......
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw7] linux -- Python 3.6.8 /usr/bin/python

self = <rucio.rse.protocols.xrootd.Default object at 0x7ff46acabf98>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
>               raise exception.RSEAccessDenied(err)
E               rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E               Details: [FATAL] Invalid address

../rse/protocols/xrootd.py:198: RSEAccessDenied

During handling of the above exception, another exception occurred:

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7ff46d4c47b8>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7ff46d4c49e8>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7ff46d479b00>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.xrootd.Default'
        scope = str(did_factory.default_scope)
        rse, _ = rse_factory.make_xroot_rse()
>       did = did_factory.upload_test_file(rse)

test_download.py:472: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
temp_factories.py:283: in upload_test_file
    self.upload_client.upload([item])
../client/uploadclient.py:237: in upload
    if rsemgr.exists(rse_settings, pfn if pfn else file_did, domain=domain, auth_token=self.auth_token, logger=logger):
../rse/rsemanager.py:279: in exists
    protocol.connect()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.xrootd.Default object at 0x7ff46acabf98>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
                raise exception.RSEAccessDenied(err)
        except Exception as e:
>           raise exception.RSEAccessDenied(e)
E           rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E           Details: Referenced RSE not reachable.
E           Details: [FATAL] Invalid address

../rse/protocols/xrootd.py:200: RSEAccessDenied
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
---------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------
WARNING  root:request.py:727 Tried to cancel non-existant request for DID mock:file_NMDZVOCFBE at RSE MOCK_YZLNLAUZCT
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_blacklisted_replicas
test_download.py::test_download_without_base_dir
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_norandom_respected
test_download.py::test_transfer_timeout
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
============================================================================== 1 failed, 7 passed, 1 skipped, 15 warnings in 12.94s ===============================================================================
[root@b2e68ef33f26 rucio]# 
``

Reply:
So, `mock://` is not a recognized GFAL protocol, as can be seen here:

`E           Details: Protocol not supported or path/url invalid: mock://3b3095674f1a4366b941e49c63e48fc9.cern.ch/test/mock/1d/04/file_HAWCDMUGBY`


For the second one:

```
E           rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E           Details: Referenced RSE not reachable.
E           Details: [FATAL] Invalid address
```

Did you start the XrootD storage server? Should come from the docker-compose `...-storage.yaml`

Reply:
@mlassnig yep
Also yes, I ran that command before entering the shell:
```
                                    dev_xrd2_1
25ff495325ce   rucio/xrootd                  "/docker-entrypoint.…"   42 hours ago   Up 42 hours   

```

Reply:
It means the hostname you are using cannot be found. From the log I see that you used `root://9dcc33d97a9b4095b9b2fab3b759b283.cern.ch/test/mock/e8/19/file_QHMZUTVJOB`... and `9dcc33d97a9b4095b9b2fab3b759b283.cern.ch` definitely does not exist in your local docker setup. So you have to check that the hostname you use there is the correct one that's also used within the docker network.

Reply:
@mlassnig @bziemons okay so, from what I can see of wherever the make_xroot_rse is used, Also for srm, both have only been used in cases of bulk adding replicas for a file. I will look into what could be causing the error

Reply:
@mlassnig Oh i see, But it is auto-generating that file along with the path, I am not interfering as such anywhere. The file is a dummy file created during upload

Reply:
It seems like the `_make_rse` wasn't made for actually uploading to the created RSE. Maybe @rcarpa can comment on the use case of it

Reply:
@bziemons wait, in file_archive the make_xroot_rse is used to create an RSE which is later used to add file replicas on that RSE. So it is being used for uploading files to the created RSE right?

Reply:
I added these functions as very simple wrappers to cover most simple initialization for test cases. They are not designed to be bullet-proof and production-ready. The protocol definition is not very customizable, and it's probably not a good idea to complicate these functions too much. In such cases, when more control over protocol customization is needed, i think it's best to call bare make_rse(). And manually add the desired protocol with rse_core.add_protocol() or rse_client.add_protocol(). See an example here https://github.com/rucio/rucio/blob/697b61ac4e6a2683ef0f814f514c74d541a17de4/lib/rucio/tests/test_archive.py#L191

Reply:
The failure or success of the upload will depend on the protocol configuration and the availability of that protocol. For example, make_posix_rse() willl result into an RSE to which you can upload. Make_xroot_rse() will not allow to upload to it; because the xroot deamons will not be actually be run; it is only a "xroot mock" which only exists as database objects. If you need to test with xroot rses, you'll have to run the test in the "integration" test suite.

Reply:
@rcarpa Thank you! So i changed it to how the example portrayed taking the protocol as xrootd, but it is still giving me the same error:
```
[root@b2e68ef33f26 rucio]# tools/pytest.sh test_download.py
Running pytest with extra arguments: test_download.py
Running pytest with pytest-xdist: -p xdist -p ruciopytest.plugin --dist=rucio --numprocesses=auto
=============================================================================================== test session starts ===============================================================================================
platform linux -- Python 3.6.8, pytest-6.1.2, py-1.10.0, pluggy-0.13.1
rootdir: /opt/rucio/lib/rucio/tests
plugins: xdist-2.2.1
gw0 [10] / gw1 [10] / gw2 [10] / gw3 [10] / gw4 [10] / gw5 [10] / gw6 [10] / gw7 [10]
s.......F
==================================================================================================== FAILURES =====================================================================================================
__________________________________________________________________________________________ test_download_file_with_impl ___________________________________________________________________________________________
[gw2] linux -- Python 3.6.8 /usr/bin/python

self = <rucio.rse.protocols.xrootd.Default object at 0x7f64cbf10be0>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
>               raise exception.RSEAccessDenied(err)
E               rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E               Details: [ERROR] Operation expired

../rse/protocols/xrootd.py:198: RSEAccessDenied

During handling of the above exception, another exception occurred:

rse_factory = <rucio.tests.temp_factories.TemporaryRSEFactory object at 0x7f64ce70b780>, did_factory = <rucio.tests.temp_factories.TemporaryDidFactory object at 0x7f64ce70b9b0>
download_client = <rucio.client.downloadclient.DownloadClient object at 0x7f64ce6c4b70>

    def test_download_file_with_impl(rse_factory, did_factory, download_client):
        impl = 'rucio.rse.protocols.gfalv2.Default'
        scope = str(did_factory.default_scope)
        rse, rseid = rse_factory.make_rse()
        add_protocol(rseid, {'scheme': 'root',
                               'hostname': 'xrootpriority1.aperture.com',
                               'port': 1410,
                               'prefix': '/prefix2/',
                               'impl': 'rucio.rse.protocols.xrootd.Default',
                               'domains': {
                                   'lan': {'read': 2, 'write': 2, 'delete': 2},
                                   'wan': {'read': 2, 'write': 2, 'delete': 2}}})
>       did = did_factory.upload_test_file(rse)

test_download.py:481: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
temp_factories.py:283: in upload_test_file
    self.upload_client.upload([item])
../client/uploadclient.py:237: in upload
    if rsemgr.exists(rse_settings, pfn if pfn else file_did, domain=domain, auth_token=self.auth_token, logger=logger):
../rse/rsemanager.py:279: in exists
    protocol.connect()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <rucio.rse.protocols.xrootd.Default object at 0x7f64cbf10be0>

    def connect(self):
        """ Establishes the actual connection to the referred RSE.
    
            :param credentials Provides information to establish a connection
                to the referred storage system. For S3 connections these are
                access_key, secretkey, host_base, host_bucket, progress_meter
                and skip_existing.
    
            :raises RSEAccessDenied
        """
        self.logger(logging.DEBUG, 'xrootd.connect: port: {}, hostname {}'.format(self.port, self.hostname))
        try:
            # The query stats call is not implemented on some xroot doors.
            # Workaround: fail, if server does not reply within 10 seconds for static config query
            cmd = 'XrdSecPROTOCOL=gsi XRD_REQUESTTIMEOUT=10 xrdfs %s:%s query config %s:%s' % (self.hostname, self.port, self.hostname, self.port)
            self.logger(logging.INFO, 'xrootd.connect: cmd: {}'.format(cmd))
            status, out, err = execute(cmd)
            if status != 0:
                raise exception.RSEAccessDenied(err)
        except Exception as e:
>           raise exception.RSEAccessDenied(e)
E           rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
E           Details: Referenced RSE not reachable.
E           Details: [ERROR] Operation expired

../rse/protocols/xrootd.py:200: RSEAccessDenied
------------------------------------------------------------------------------------------------ Captured log call ------------------------------------------------------------------------------------------------
WARNING  root:uploadclient.py:368 Scope mock not found for the account root.
---------------------------------------------------------------------------------------------- Captured log teardown ----------------------------------------------------------------------------------------------
WARNING  root:request.py:727 Tried to cancel non-existant request for DID mock:file_KKFZCBMLQE at RSE MOCK_VCLLAIDSUR
================================================================================================ warnings summary =================================================================================================
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
../client/baseclient.py:55
  /opt/rucio/lib/rucio/client/baseclient.py:55: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    import imp

test_download.py::test_download_blacklisted_replicas
test_download.py::test_download_without_base_dir
test_download.py::test_norandom_respected
test_download.py::test_transfer_timeout
test_download.py::test_trace_copy_out_and_checksum_validation
test_download.py::test_download_to_two_paths
test_download.py::test_download_multiple
  /usr/local/lib/python3.6/site-packages/urllib3/connectionpool.py:1020: InsecureRequestWarning: Unverified HTTPS request is being made to host 'rucio'. Adding certificate verification is strongly advised. See: https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
    InsecureRequestWarning,

-- Docs: https://docs.pytest.org/en/stable/warnings.html
============================================================================================= short test summary info =============================================================================================
FAILED test_download.py::test_download_file_with_impl - rucio.common.exception.RSEAccessDenied: Referenced RSE not reachable.
============================================================================== 1 failed, 7 passed, 1 skipped, 15 warnings in 24.96s ===============================================================================
[root@b2e68ef33f26 rucio]# 

```

From the looks of it, the upload_test_file seems to not work with this no matter what. In this situation and all places where xrootd and srm protocols is used, its used with only add_replicas which in turn deals directly with the DB at the RSE 
On the contrary, upload_test_file calls the upload function and tries actually connecting using the protocol which gives this error of RSE access denied. 
(please correct me if I am mistaken anywhere)

Reply:
> The failure or success of the upload will depend on the protocol configuration and the availability of that protocol. For example, make_posix_rse() willl result into an RSE to which you can upload. Make_xroot_rse() will not allow to upload to it; because the xroot deamons will not be actually be run; it is only a "xroot mock" which only exists as database objects. If you need to test with xroot rses, you'll have to run the test in the "integration" test suite.

thank you, I just got this.